File: Parsers/parser.js

/*la fonction parseOut constitue le message qui serra envoyé vers
l'application, en fonction d'une action definie en conception et
des parametre dont cette action a besoin */

/*le parametre est un objet de type {URL:, login:, mdp:,nomFieldLogin:,nomFieldMdp:}*/

function parseOut(action, parametre){

//declaration de variable correspondant aux mots cles de notre protocole
	var add = "INSERT:";
	var read = "SELECT:";
	var suppr = "DELETE:";
	var maj	= "UPDATE:";

	var query = "Q:";
	var anser = "A:";
	var noanser = "X:";

	var modF = "ModifForm";
	var modI = "ModifID";
	var addS = "AjoutSite";
	var fillF = "RemplirForm";
	var supp = "SupprID";
	var addI = "AjoutID";
	var chkI = "CheckID";


	var toSend;

//parsing en fonction de l'action
	switch(action){
		case modF:
			toSend = noanser+maj+action+":"+parametre.nomFieldLogin+";"+parametre.nomFieldMdp+";"+parametre.URL;
			break;
		case modI:
			toSend = noanser+maj+action+":"+parametre.login+";"+parametre.mdp+";"+parametre.URL;
			break;
		case addS:
			toSend = noanser+add+action+":"+parametre.nomFieldLogin+";"+parametre.nomFieldMdp+";"+parametre.URL;
			break;
		case fillF:
			toSend = query+read+action+":"+parametre.nomFieldLogin+";"+parametre.nomFieldMdp+";"+parametre.URL;
			break;
		case supp:
			toSend = noanser+suppr+action+":"+parametre.URL;
			break;
		case addI:
			toSend = noanser+add+action+":"+parametre.login+";"+parametre.mdp+";"+parametre.URL;
			break;
		case chkI:
			toSend = query+read+action+":"+parametre.login+";"+parametre.mdp+";"+parametre.URL;
			break;
		default:
			toSend = "erreur ParseOut: action inconnue ";
	}


	return Base64.encode(toSend);
}


/* la fonction parseIn decortique les messages arrivant a l'add-on
depuis l'application, ce sont des messages contenant des liste d'id*/

function parseIn(messageEncoded){

	var message = Base64.decode(messageEncoded);

	var separateur = ":"
	var separateurParam = "|";
	var anser = "A";


	var messageSplit = message.split(separateur);
	var log;
	var logSplit;
	var logArr = [] ;

	var requeteMdp = "query";
	var requeteRead = "exist";
	if(messageSplit[0]==anser){
		if(messageSplit[1]==requeteMdp){
			for(var i=2; i<messageSplit.length ; i++){
				logSplit=messageSplit[i].split(separateurParam);
				log={login:logSplit[0], mdp:logSplit[1]};
				logArr.push(log);
			}
		} else if (messageSplit[1] == requeteRead) {
			if(messageSplit[2] == "true")
				return true;
			else if (messageSplit[2] == "false")
				return false;
			else
			  return "erreur: ParserIn js, reponse non reconnue";
		}else {
			return "erreur: ParserIn js, combinanison fausse";
		}
	}
	else{
		return "erreur: parseIn js";
	}
	return logArr; //retourne un tableau d'objet de type {login:,mdp:}
}


var Base64={_keyStr:"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",
							encode:function(e){
								var t="";
								var n,r,i,s,o,u,a;
								var f=0;
								e=Base64._utf8_encode(e);
								while(f<e.length){
									n=e.charCodeAt(f++);
									r=e.charCodeAt(f++);
									i=e.charCodeAt(f++);
									s=n>>2;
									o=(n&3)<<4|r>>4;
									u=(r&15)<<2|i>>6;
									a=i&63;
									if(isNaN(r)){
										u=a=64
									}
									else if(isNaN(i)){
											a=64
									}
									t=t+this._keyStr.charAt(s)+this._keyStr.charAt(o)+this._keyStr.charAt(u)+this._keyStr.charAt(a)
								}
								return t
							},
							decode:function(e){
								var t="";
								var n,r,i;
								var s,o,u,a;
								var f=0;
								e=e.replace(/[^A-Za-z0-9\+\/\=]/g,"");
								while(f<e.length){
									s=this._keyStr.indexOf(e.charAt(f++));
									o=this._keyStr.indexOf(e.charAt(f++));
									u=this._keyStr.indexOf(e.charAt(f++));
									a=this._keyStr.indexOf(e.charAt(f++));
									n=s<<2|o>>4;r=(o&15)<<4|u>>2;
									i=(u&3)<<6|a;
									t=t+String.fromCharCode(n);
									if(u!=64){
										t=t+String.fromCharCode(r);
									}
									if(a!=64){
										t=t+String.fromCharCode(i);
									}
								}
								t=Base64._utf8_decode(t);
								return t;
							},
							_utf8_encode:function(e){
								e=e.replace(/\r\n/g,"\n");
								var t="";
								for(var n=0;n<e.length;n++){
									var r=e.charCodeAt(n);
									if(r<128){
										t+=String.fromCharCode(r);
									}else if(r>127&&r<2048){
										t+=String.fromCharCode(r>>6|192);
										t+=String.fromCharCode(r&63|128);
									}
									else{
										t+=String.fromCharCode(r>>12|224);
										t+=String.fromCharCode(r>>6&63|128);
										t+=String.fromCharCode(r&63|128);
									}
								}
								return t;
							},
							_utf8_decode:function(e){
								var t="";
								var n=0;
								var r=c1=c2=0;
								while(n<e.length){
									r=e.charCodeAt(n);
									if(r<128){
										t+=String.fromCharCode(r);
										n++;
									}
									else if(r>191&&r<224){
										c2=e.charCodeAt(n+1);
										t+=String.fromCharCode((r&31)<<6|c2&63);
										n+=2;
									}
									else{
										c2=e.charCodeAt(n+1);
										c3=e.charCodeAt(n+2);
										t+=String.fromCharCode((r&15)<<12|(c2&63)<<6|c3&63);
										n+=3;
									}
								}
								return t;
							}
							}

File: JAVA/Libraries/SrcLibraries/src/initToken_old/Parser.java
package initToken;

import org.apache.commons.codec.binary.Base64;

public class Parser {
	
	
	public Parser(){
		
	}
	
	//macro
	public static final int ACTION_ID = 1;
	public static final int ACTION_PARAM_SITE = 2;
	public static final int ACTION_SUPPR = 3;
	
	public static final String MODF = "ModifForm";
	public static final String MODI = "ModifID";
	public static final String ADDS = "AjoutSite";
	public static final String ADDI = "AjoutID";
	public static final String FILLF = "RemplirForm";
	public static final String SUPP = "SupprID";
	public static final String CHKI = "CheckID";
	
	public static final String ACTION_ADD = "INSERT";
	public static final String ACTION_MAJ = "UPDATE";
	public static final String ACTION_DEL = "DELETE";
	public static final String ACTION_SEL ="SELECT";
	
	public static final String SEP_MSG = ":";
	public static final String SEP_PARAM = ";";
	public static final String SEP_ANSER = "|";
	
	public static final String QUERY = "Q";
	public static final String ANSER = "A";
	public static final String NO_ANSER = "X";
	
	public static final String REQUETE_MDP = "query";
	public static final String REQUETE_EXIST = "exist";
	
	
	private String actionType;
	private String action;
	private String URL;
	private String login;
	private String mdp;
	private String fieldLogin;
	private String fieldMdp;
	private boolean anser;
	
	//parse les message sortant contenant les mdp
	public String ParseOut(String[] login, String[] mdp){
		StringBuffer toSend = new StringBuffer();
		
		toSend.append(ANSER); //les messages retournés a l'add-on commence par A (anser) 
		toSend.append(SEP_MSG);
		toSend.append(REQUETE_MDP);
		
		
		if(login.length == mdp.length){
			for (int i = 0; i < mdp.length; i++) {
				toSend.append(SEP_MSG);//séparateur des différents champs 
				toSend.append(login[i]);
				toSend.append(SEP_ANSER);  //separateur pour les parametres (ici les mdp)
 				toSend.append(mdp[i]);
			}
		}
		else{
			return "erreurParseOut";
		}
		
		byte [] encodedToSend = Base64.encodeBase64(toSend.toString().getBytes());
		
		return new String(encodedToSend);
	}
	
	public String ParseOutExist(boolean anser){
		StringBuffer toSend = new StringBuffer();
		
		toSend.append(ANSER);
		toSend.append(SEP_MSG);
		toSend.append(REQUETE_EXIST);
		toSend.append(SEP_MSG);
		toSend.append(anser);
		
		byte [] encodedToSend = Base64.encodeBase64(toSend.toString().getBytes());
		
		return new String (encodedToSend);
	}
	
	
	
	
	//parser pour les messages entrant	
	public void ParseIn(String messageEncoded){
		byte [] decoded = Base64.decodeBase64(messageEncoded);
		String message = new String(decoded);
		
		
		
		flushAll();
		
		String [] msgSplit = message.split(SEP_MSG,4); //parsing selon le separateur :
		
		switch (msgSplit[0]) {
		case QUERY: // requete necessitant une reponse
			anser=true;
			break;
		case NO_ANSER: // requete sans reponse
			anser=false;
			break;
		default:
			System.out.println("erreur ParseIn: message incorrecte");
			return;
		}
		
		switch (msgSplit[1]) {
		case ACTION_ADD: //requete d'insertion pour le SQL
			actionType = msgSplit[1];
			switch (msgSplit[2]) {
			case ADDI: //requete d'ajout identifiant  
				action=msgSplit[2];
				parseParam(msgSplit[3], ACTION_ID);
				//inserer fonction de traitement
				break;

			case ADDS: // requete ajour d'un nouveau site
				action=msgSplit[2];
				parseParam(msgSplit[3], ACTION_PARAM_SITE);
				//inserer fonction de traitement
				break;
				
			default:
				System.out.println("erreur ParseIn: combinaison fausse1");
				return;
			}
			
			break;
		
		case ACTION_SEL: //requete pour recuperer des donnée sur le token
			actionType = msgSplit[1];
			switch (msgSplit[2]) {
			case FILLF: // recupération mdp pour le formulaire
				action=msgSplit[2];
				parseParam(msgSplit[3], ACTION_PARAM_SITE);
				//inserer fonction de traitement
				break;
				
			case CHKI:
				action = msgSplit[2];
					parseParam(msgSplit[3], ACTION_ID);
					

			default:
				System.out.println("erreur ParseIn: combinaison fausse2");
				return;
			}
				
			break;
		
		case ACTION_MAJ: // requetes de mise a jour pour les données
			actionType = msgSplit[1];
			
			switch (msgSplit[2]) {
			case MODF: // mise a jour des info sur le formulaire
				action=msgSplit[2];
				parseParam(msgSplit[3], ACTION_PARAM_SITE);
				//inserer fonction de traitement ou autre proposition
				break;
			
			case MODI: //mise a jour d'identifiant utilisateur
				action=msgSplit[2];
				parseParam(msgSplit[3], ACTION_ID);
				//inserer fonction traitement
				break;
				
			default:
				System.out.println("erreur ParseIn: combinaison fausse3");
				return;
			}
			
			
			break;
			
		case ACTION_DEL: //suppresion de données
			actionType = msgSplit[1];
			
			switch (msgSplit[2]) {
			case SUPP: // supression  d'identifiant utilisateur
				action=msgSplit[2];
				parseParam(msgSplit[3], ACTION_SUPPR);
				//inserser fonction de traitement
				break;

			default:
				System.out.println("erreur ParseIn: combinaison fausse4");
				return;
			}
			
			break;
			
		default:
			System.out.println("Erreur ParseIn: actionType inconnue ");
			return;
		}
		
	
		
	}
	
	// parsing du champs de parametre de 3 façons différentes
	
	public void parseParam(String param , int action){
		
		String [] paramSplit = param.split(SEP_PARAM); // parsing du champs en focntion du separateur ';'
		
		
		switch(action){
		case ACTION_ID:  //ajout mdp ou modif logs
			
			login = paramSplit[0];
			mdp = paramSplit[1];
			URL = paramSplit[2];
			break;
		case ACTION_PARAM_SITE: //changement champ ou ajout site ou remplir form
			URL = paramSplit[2];
			fieldLogin=paramSplit[0];
			fieldMdp=paramSplit[1];
			break;
		case ACTION_SUPPR: //suppression url
			URL=paramSplit[0];
			break;
		default :
			System.out.println("erreur parser argument");
						
		
		}
	
			
	}
	
	
	//cette fonction remet tous les 
	private void flushAll(){
		action=null;
		actionType=null;
		URL=null;
		login=null;
		mdp=null;
		fieldLogin=null;
		fieldMdp=null;
		anser=false;
	}
	

	public boolean getAnser() {
		return anser;
	}

	public String getAction(){
		return action;
	}
	
	public String getActionType(){
		return actionType;
	}
	
	public String getURL(){
		return URL;
	}
	
	public String getLogin(){
		return login;
	}
	
	public String getMdp(){
		return mdp;
	}

	public String getFieldLogin(){
		return fieldLogin;
	}
	
	public String getFieldMdp(){
		return fieldMdp;
	}
}
